File: java/parser/src/main/java/gov/lbl/parser/parser/FactorVisitor.java
package gov.lbl.parser.parser;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Factor;
import gov.lbl.parser.domain.Primary;

import static java.util.stream.Collectors.toList;

import java.util.List;

public class FactorVisitor extends modelicaBaseVisitor<Factor> {
    @Override
    public Factor visitFactor(modelicaParser.FactorContext ctx) {
        PrimaryVisitor primaryVisitor = new PrimaryVisitor();
        List<Primary> primarys = ctx.primary() == null ? null : ctx.primary()
                                                                .stream()
                                                                .map(primary -> primary.accept(primaryVisitor))
                                                                .collect(toList());
        
        String op = 
                ctx.SYMBOL_CARET()==null ? ctx.SYMBOL_DOTCARET().getText() : ctx.SYMBOL_CARET().getText();
        
        Primary primary1 = primarys.get(0);
        Primary primary2 = primarys.get(1);
        
        return new Factor(primary1, op, primary2);
    }
}

File: jsParser/parser/FactorVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Factor = require('../domain/Factor')

const PrimaryVisitor = require('./PrimaryVisitor')

class FactorVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitFactor (ctx) {
    const op = ctx.SYMBOL_CARET() ? ctx.SYMBOL_CARET().getText() : (ctx.SYMBOL_DOTCARET() ? ctx.SYMBOL_DOTCARET().getText() : '')
    const primarys = []
    let primary1 = null
    let primary2 = null

    if (ctx.primary()) {
      const primaryVisitor = new PrimaryVisitor.PrimaryVisitor()
      ctx.primary().forEach(pri => {
        primarys.push(primaryVisitor.visitPrimary(pri))
      })
    }

    if (primarys.length > 0) {
      primary1 = primarys[0]
      if (primarys.length > 1) {
        primary2 = primarys[1]
      }
    }
    return new Factor.Factor(primary1, op, primary2)
  }
}

FactorVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitFactor = this.visitFactor
exports.FactorVisitor = FactorVisitor
